docs/design/variadics.md

```diff
@@ -0,0 +1,1018 @@
+    equals `C` given a binding map `M1`, and `B` equals `D` given a binding map
+    `M2`, and `M1` ∪ `M2` is consistent, then `P(A, B)` equals `P(C, D)` given a
+    binding map `M1` ∪ `M2`.
+
+Almost all of the deduction rules can be generalized in the same way: the
+binding map in the conclusion is the union of the binding maps in the premises,
+and there's an additional premise that this union is consistent. If there are no
+"equals" or "convertible to" relations in the premises, the binding map in the
+conclusion is empty. When that pattern applies, we will usually leave the
+binding maps implicit (as in the original formulation of the above rule). One
+major exception is the following rule:
+
+_Binding introduction:_ Let `X` be a deduced parameter with type `T`. If the
+type of `E` is convertible to `T` given binding map `M`, and `M` ∪ {(`X`, `E`)}
+is consistent, then `X` equals `E` given a binding map `M` ∪ {(`X`, `E`)}.
+
+The deduction rules that are specific to variadic types are as follows:
+
+-   _Expansion convertibility:_ `...T` is convertible to `...U` if the arity of
+    `U` equals the arity of `T`, and all scalar components of `T` are
+    convertible to all scalar components of `U`.
+-   _Shape equality:_ Let `(S1s)`, `(S2s)`, `(S3s)`, and `(S4s)` be shapes.
+    `(S1s, S2s)` equals `(S3s, S4s)` if `(S1s)` equals `(S3s)` and `(S2s)`
+    equals `(S4s)`.
+
+#### Example
+
+Suppose we are typechecking the following function call:
+
+```carbon
+fn F[T:! type](x: Vector(T));
+F(() as Vector(i32));
+```
+
+We need to verify that `Vector(i32)` is convertible to `Vector(T)`. We can
+construct a proof by applying the above inference rules, as follows:
+
+1. (By identity) `type` equals `type` given a binding map {}.
+2. (By equality implies convertibility) `type` is convertible to `type` given a
+   binding map {}.
+3. (By binding introduction) `T` equals `i32` given a binding map {(`T`,
+   `i32`)}.
+4. (By parameterized type equality) `Vector(T)` equals `Vector(i32)` given a
+   binding map {(`T`, `i32`)}.
+5. (By equality implies convertibility) `Vector(T)` is convertible to
+   `Vector(i32)` given a binding map {(`T`, `i32`)}.
+
+Notice that as a byproduct, this proof gives us a binding map with the deduced
+values of the parameters.
+
+Conceptually, the typechecker constructs this proof by searching backwards from
+desired conclusion without considering binding maps, and then running the
+resulting proof forward to compute the final binding map.
+
+### Deduction algorithm
+
+A full pattern is in _normal form_ if it contains no pack literals, and every
+arity coercion is fully expanded. Note that all user-written full patterns are
+in normal form. Note also that by construction, this means that the type of the
+body of every pack expansion has a single scalar component. The _canonical form_
+of a full pattern is the unique normal form (if any) that is "maximally merged",
```

Comment:
Naively "maximally merged" sounds like the opposite of "arity coercion is fully expanded"